File: output.jl
function print_edge(rg_out, edge, mean_aff)
    area = edge.num
    sum_aff = edge.sum
    if mean_aff > 0
        sum_aff = area*mean_aff
    else
        sum_aff *= 0.9995
    end
    p = minmax(edge.p1, edge.p2)
    push!(rg_out,"$(p[1]) $(p[2]) $(sum_aff) $(area) $(edge.v1) $(edge.v2) $(edge.aff) $(edge.area)\n")
end

function update_new_rg(num_seg, new_rg, matches)
    f = open("axon.in","w")
    visited = Set{atomic_edge}()
    rg_out = AbstractString[]
    for edge in keys(matches)
        print_edge(rg_out, edge, matches[edge])
        push!(visited, edge)
    end
    for a in keys(new_rg)
        for b in keys(new_rg[a])
            edge = new_rg[a][b]
            if edge in visited
                continue
            else
                print_edge(rg_out, edge, -1)
                push!(visited, edge)
            end
        end
    end
    write(f,"$(num_seg-1) $(num_seg) $(length(rg_out))\n")
    for str in rg_out
        write(f, str)
    end
    close(f)
end

function merge_edges(merge_graph, threshold, new_rg)
    edges = OrderedDict{atomic_edge, Float64}()
    visited = Set{Int}()
    for p in keys(merge_graph)
        axons_group = Int[]
        if p in visited
            continue
        end
        queue = Queue(Int)
        enqueue!(queue, p)
        push!(axons_group, p)

        while length(queue) > 0
            root = dequeue!(queue)
            if root in visited
                continue
            end

            push!(visited,root)
            for neighboor in merge_graph[root]
                if !(neighboor in visited)
                    enqueue!(queue, neighboor)
                    push!(axons_group, neighboor)
                end
            end
        end
        for q in combinations(axons_group,2)
            if haskey(new_rg[q[1]], q[2])
                edges[new_rg[q[1]][q[2]]] = threshold
            end
        end
    end
    return edges
end

File: src/Output.cpp
#include "Output.h"
#include <QQueue>
#include <QFile>
#include <QStringList>
#include <QtDebug>

void OutputRegionGraph::appendEdges(const SupervoxelDict & mergeGraph, value_type threshold)
{
    SupervoxelSet visited;
    qDebug() << "Merge" << mergeGraph.keys().size() << "segments at threshold:" << threshold ;
    foreach (auto p, mergeGraph.keys()) {
        QVector<id_type> segGroup;
        if (visited.contains(p)) {
            continue;
        }
        QQueue<id_type> queue;
        queue.enqueue(p);
        segGroup.append(p);
        while (!queue.isEmpty()) {
            auto root = queue.dequeue();
            if (visited.contains(root)) {
                continue;
            }
            visited.insert(root);
            foreach(auto neighbour, mergeGraph[root]) {
                if (!visited.contains(neighbour)) {
                    queue.enqueue(neighbour);
                    segGroup.append(neighbour);
                }
            }
        }
        for (int i = 0; i < segGroup.size(); i++) {
            for (int j = i+1; j < segGroup.size(); j++) {
                if (m_regionGraph[segGroup[i]].contains(segGroup[j])) {
                    m_edges[m_regionGraph[segGroup[i]][segGroup[j]]] = threshold;
                }
            }
        }
    }
}

QString OutputRegionGraph::printEdge(const MeanPlusEdge * edge, value_type meanAffinity)
{
    auto area = edge->num;
    auto sum_aff = edge->sum;
    if (meanAffinity > 0) {
        sum_aff = area*meanAffinity;
    } else {
        sum_aff *= 0.9995;
    }
    return QString("%1 %2 %3 %4 %5 %6 %7 %8\n").arg(edge->p1).arg(edge->p2).arg(sum_aff).arg(area).arg(edge->v1).arg(edge->v2).arg(edge->aff).arg(edge->area);
}

void OutputRegionGraph::updateRegionGraph(id_type maxSegId)
{
    QSet<const MeanPlusEdge *> visited;
    QStringList region_graph_output;
    foreach (auto edge, m_edges.keys()) {
        region_graph_output << printEdge(edge, m_edges[edge]);
        visited.insert(edge);
    }
    foreach (auto a, m_regionGraph.keys()) {
        foreach (auto b, m_regionGraph[a].keys()) {
            auto edge = m_regionGraph[a][b];
            if (visited.contains(edge)) {
                continue;
            }
            region_graph_output << printEdge(edge, -1);
            visited.insert(edge);
        }
    }
    QFile outputFile("axon.in");
    outputFile.open(QIODevice::WriteOnly | QIODevice::Text);
    if(!outputFile.isOpen()){
        qDebug() << "- Error, unable to open" << "axon.in" << "for output";
        return;
    }
    QTextStream outputStream(&outputFile);
    outputStream << maxSegId << " " << maxSegId+1 << " "<< region_graph_output.size() << "\n";
    foreach (auto s, region_graph_output) {
        outputStream << s;
    }
}
